### PR TITLE
OrangePi i96 initial support

### PR DESCRIPTION
**Pull request by daniel-thompson**
Initial support is based on the vendor kernel and u-boot.

The vendor code currently has minimal changes, however they are modified to support gcc-7.

### CODE REVIEW

File: conf/machine/orangepi-i96.conf [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+#@DESCRIPTION: OrangePi i96 machine
+
+require conf/machine/include/tune-cortexa5.inc
+
+PREFERRED_PROVIDER_virtual/xserver ?= "xserver-xorg"
+
+XSERVER ?= "xserver-xorg \
+            mesa-driver-swrast \
+            xf86-input-evdev \
+            xf86-video-fbdev"
```
Comment (fboudra) [inline conf/machine/orangepi-i96.conf:12]:
```
            xf86-video-fbdev \
"
```

File: conf/machine/orangepi-i96.conf [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+KERNEL_IMAGETYPE = "zImage"
+#KERNEL_DEVICETREE = "currently-uses-a-concatenated-dt"
+
+UBOOT_MACHINE = "rda8810_config"
+UBOOT_SUFFIX = "rda"
+
+# TODO: Switch this to LS-UART1 and reduce to 115200 (921600 is needed
+#       on ttyS0 in order to match speeds with the boot ROM output
+SERIAL_CONSOLE = "921600 ttyS0"
```
Comment (fboudra) [inline conf/machine/orangepi-i96.conf:26]: The SERIAL_CONSOLE variable is deprecated. Please use the SERIAL_CONSOLES variable. The format is slightly different, it isn't only a renaming of the variable.

File: conf/machine/orangepi-i96.conf [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+MACHINE_ESSENTIAL_EXTRA_RDEPENDS += "\
+    u-boot-orangepi-i96 \
+"
+
+CMDLINE_ROOT_EMMC   = "mmcblk0p2"
```
Comment (fboudra) [inline conf/machine/orangepi-i96.conf:32]: only 1 space: CMDLINE_ROOT_EMMC = "mmcblk0p2"

File: recipes-bsp/u-boot/u-boot-orangepi-i96.bb [hunk 4]
```diff
@@ -0,0 +1,35 @@
+require ${COREBASE}/meta/recipes-bsp/u-boot/u-boot.inc
+
+FILESEXTRAPATHS_prepend := "${THISDIR}/u-boot:"
```
Comment (fboudra) [inline recipes-bsp/u-boot/u-boot-orangepi-i96.bb:3]: we don't apply any patches or ship extra files, so the line can be removed entirely.

File: recipes-bsp/u-boot/u-boot-orangepi-i96.bb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+SUMMARY = "U-Boot bootloader for OrangePi i96"
```
Comment (fboudra) [inline recipes-bsp/u-boot/u-boot-orangepi-i96.bb:5]: the file doesn't follow OE style guidelines. See u-boot-juno and u-boot-poplar that have been reformatted yesterday.

File: recipes-kernel/linux/linux-orangepi-i96_git.bb [hunk 6]
```diff
@@ -0,0 +1,58 @@
+require linux.inc
+
+DESCRIPTION = "OrangePi i96 vendor kernel"
+
+PV = "3.10.62+git${SRCPV}"
+SRCREV_kernel = "b476a2c5bfb893aba3ee4daa840196a97898bf48"
+SRCREV_FORMAT = "kernel"
+
+SRC_URI = "git://github.com/daniel-thompson/linux.git;nobranch=1;name=kernel"
```
Comment (fboudra) [inline recipes-kernel/linux/linux-orangepi-i96_git.bb:9]: can you add also protocol=https

File: recipes-kernel/linux/linux-orangepi-i96_git.bb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+S = "${WORKDIR}/git"
+
+COMPATIBLE_MACHINE = "orangepi-i96"
+KERNEL_IMAGETYPE ?= "zImage"
+
+# make[3]: *** [scripts/extract-cert] Error 1
+DEPENDS += "openssl-native"
+HOST_EXTRACFLAGS += "-I${STAGING_INCDIR_NATIVE}"
+
+#LDFLAGS += "${TOOLCHAIN_OPTIONS}"
```
Comment (fboudra) [inline recipes-kernel/linux/linux-orangepi-i96_git.bb:20]: if not used, get rid of it.

**Review by fboudra — COMMENTED** [review]
There's a few minor nitpicks, but overall it LGTM.

**Comment by daniel-thompson** [comment]
Thanks Fathi. Will make all the proposed changes. I'll try to get things updates in the next day or three.

File: conf/machine/orangepi-i96.conf (hunk 1 above)
Comment (koenkooi) [reply conf/machine/orangepi-i96.conf:12]: Can we change 'fbdev' to 'modesetting'?

**Comment by parrotgeek1** [comment]
Update?

**Comment by daniel-thompson** [comment]
@parrotgeek1 : Acting on the review comments is still on my TODO list but has got pretty deeply buried by other things. It will be a while longer before I can repost.

**Comment by fboudra** [comment]
@daniel-thompson based on the minor fixes that has been pointed at the review. we/I can take it from there and make the fixes myself before merging.

File: conf/machine/orangepi-i96.conf (hunk 1 above)
Comment (daniel-thompson) [reply conf/machine/orangepi-i96.conf:12]: This thing is running with a v3.10 vendor kernel. I haven't checked explicitly (i96 not having easy access to video out) but I'd be surprised if it has a working KMS implemention.

**Comment by daniel-thompson** [comment]
@fboudra : A generous offer if you have the time... (I've just replied to Koen's comment about fbdev since I think that needs more then OE knowhow to conclude on).

**Comment by parrotgeek1** [comment]
@daniel-thompson https://github.com/RDA8810/linux-RDA8810
is done by me based on linaro android 3.10.62 kernel 
(which results in only a few hundred files changing, not thousands like you have)

**Comment by daniel-thompson** [comment]
@parrotgeek1:
So my kernel is nothing more or less than the vendor kernel with a script run over it (and some gcc-7 patches). It's based on 3.10.62 because I hoped to experiment with a technique for ripping out all the Android stuff... If you reorder the patches so the board support is at the front of the patch queue and all the Android support is at the back and then start testing at the *last* board support patch then all the remaining Android patches are still there to bisect through if you find driver regressions that really do rely on some of the non-BSP patches.

However, it doesn't look like I'll get round to that anytime soon so perhaps I should regenerate everything from a linaro-android baseline instead.

**Comment by parrotgeek1** [comment]
@daniel-thompson Almost all of the patches were either backported security fixes, or odd attempts at decreasing RAM usage. I kept the ones that weren't.

**Comment by parrotgeek1** [comment]
I can't edit comments on my phone, but I wanted to add that I chose my way so that I could tell which of the Android related changes were made by Google or RDA

**Comment by fboudra** [comment]
Merged:
* https://github.com/96boards/meta-96boards/commit/2420c76abbc4e9e61fb31862f3122f0fc0bb1d90
* https://github.com/96boards/meta-96boards/commit/1d985cdfe8eff4ad9b729c666eea4a40eb04bec1
* https://github.com/96boards/meta-96boards/commit/3aacae0c373fdd454a723cf48e68cea644b00ada

Thanks Daniel!

**Pull request closed without merging**